# Kalign3
[Kalign3](https://github.com/TimoLassmann/kalign) is a fast multiple sequence alignment program for biological sequences.

## How to Access
Click Here: [https://console.latch.bio/explore/81603/info](https://console.latch.bio/explore/81603/info)

## Input Options
1. Fasta File: A concatenated `fasta file` containing all the sequences to be aligned.
2. Output Format: The `output format` of the alignment. Default is `CLUSTAL`.
3. Penalties: `gap open penalty`, `gap extension penalty` and `terminal gap penalties`.

## Documentation
Visit Here: [https://github.com/TimoLassmann/kalign](https://github.com/TimoLassmann/kalign)

## Benchmark results
Here are some benchmark results 

### Balibase

![Balibase_scores](https://user-images.githubusercontent.com/8110320/66697423-7ea3d000-eca3-11e9-919a-995ca8e9f7c1.jpeg)

### Bralibase

![Bralibase_scores](https://user-images.githubusercontent.com/8110320/66697424-86637480-eca3-11e9-90ea-238f82b0ac6b.jpeg)

### Homfam

![Homfam_scores](https://user-images.githubusercontent.com/8110320/66697425-895e6500-eca3-11e9-97e7-63f3a79133cf.jpeg)

### Quantest2

![Quantest2_scores](https://user-images.githubusercontent.com/8110320/66698153-6c2c9500-eca9-11e9-904c-3d6ea9a1c44d.jpeg)

## Please cite:
1. Lassmann, Timo. _Kalign 3: multiple sequence alignment of large data sets._ **Bioinformatics** (2019). [pdf](https://academic.oup.com/bioinformatics/advance-article-pdf/doi/10.1093/bioinformatics/btz795/30314127/btz795.pdf)
